pkg/protocol/http/http.go

```diff
@@ -146,46 +102,104 @@ type ClientBuilder struct {
 	supportHTTPS bool
 }
 
-func (cb *ClientBuilder) NewProtocolClient(clientOpt interface{}) (protocol.Client, error) {
-	var (
-		transport = DefaultTransport
-		tlsConfig *tls.Config
-	)
+// WithTransport allows to set transport for http protocol of  protocol.Client.
+func WithTransport(transport *http.Transport) func(protocol.Client) {
+	return func(client protocol.Client) {
+		cli := client.(*Client)
+		cli.transport = transport
+	}
+}
 
-	if clientOpt != nil {
-		opt, ok := clientOpt.(*ClientOpt)
-		if !ok {
-			return nil, errortypes.ErrConvertFailed
+// WithTLS allows to set tls config for http protocol of  protocol.Client.
+func WithTLS(config *tls.Config) func(protocol.Client) error {
+	return func(client protocol.Client) error {
+		cli := client.(*Client)
+		if cli.transport == DefaultTransport {
+			cli.transport = newDefaultTransport()
 		}
 
-		tran := opt.Get(HTTPTransport)
-		if tran != nil {
-			transport = tran.(*http.Transport)
+		cli.transport.TLSClientConfig = config
+		return nil
+	}
+}
+
+// WithMapString allows to set some options by map string.
+func WithMapString(opt map[string]string) func(client protocol.Client) error {
+	return func(client protocol.Client) error {
+		return nil
+	}
+}
+
+// WithMapInterface allows to set some options by map interface.
+// Supported:
+// key: "tls.config", value: *tls.Config
+// key: "http.transport" value: *http.Transport
+func WithMapInterface(opt map[string]interface{}) func(client protocol.Client) error {
+	return func(client protocol.Client) error {
+		var (
+			transport *http.Transport
+			config    *tls.Config
+			ok        bool
+		)
+
+		cli := client.(*Client)
+		for k, v := range opt {
+			switch k {
+			case "tls.config":
+				config, ok = v.(*tls.Config)
+				if !ok {
+					return errortypes.ErrConvertFailed
+				}
+			case "http.transport":
+				transport, ok = v.(*http.Transport)
+				if !ok {
+					return errortypes.ErrConvertFailed
+				}
+			default:
+			}
 		}
 
-		config := opt.Get(TLSConfig)
-		if config != nil {
-			tlsConfig = config.(*tls.Config)
+		if transport == nil {
+			transport = DefaultTransport
 		}
 
-		// set tls config to transport
 		if config != nil {
 			if transport == DefaultTransport {
 				transport = newDefaultTransport()
 			}
 
-			transport.TLSClientConfig = tlsConfig
+			transport.TLSClientConfig = config
+		}
+
+		cli.transport = transport
+		cli.client = &http.Client{
+			Transport: transport,
 		}
+
+		return nil
+	}
+}
+
+func (cb *ClientBuilder) NewProtocolClient(opts ...func(client protocol.Client) error) (protocol.Client, error) {
```

Comment:
Do we still need ClientBuilder? What're the extra value added by using a builder?

Reply:
ClientBuilder is to generate client for protocol client.  For some protocol, the client needs to set some specific info to initialize client by default.